### PR TITLE
Added configurator wizard in template and variant

### PR DESCRIPTION
**Pull request by asifdhuka**

### CODE REVIEW

File: configurator_wizard_in_product/models/product.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        cfg_steps = self.config_step_line_ids
+        active_step = str(cfg_steps[0].id) if cfg_steps else 'configure'
+
+        wizard_obj = self.env['product.configurator']
+        wizard = wizard_obj.create({
+            'product_tmpl_id': self.id,
+            'state': active_step,
+        })
+        ctx_no_type = self._context.copy()
+        ctx_no_type.pop('default_type', None)
```
Comment (coleste) [inline configurator_wizard_in_product/models/product.py]: comment why, better to use del ?

File: configurator_wizard_in_product/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+
+from . import models
+from . import wizard
```
Comment (coleste) [inline configurator_wizard_in_product/__init__.py]: newline

File: configurator_wizard_in_product/views/product_views.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    </record>
+
+    <record id="product_template_only_form_view" model="ir.ui.view">
+        <field name="name">product.template.product.form</field>
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_only_form_view"/>
+        <field name="arch" type="xml">
+            <xpath expr="//div[@name='options']" position="inside">
+                <div>
+                    <button name="action_wizard_product_template_configurator"
```
Comment (coleste) [inline configurator_wizard_in_product/views/product_views.xml]: better function name plz

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                raise ValidationError(
+                    _('Invalid configuration! Please check all '
+                      'required steps and fields.')
+                )
+
+        if self.env.context.get('active_model') == 'purchase.order':
+            order = self.env['purchase.order'].browse(self.env.context.get('active_id'))
+            line_vals = {'product_id': variant.id, 'date_planned': datetime.today().strftime(DEFAULT_SERVER_DATETIME_FORMAT),
+                         'product_uom': uom, 'price_unit': price, 'product_qty': 1}
+        elif self.env.context.get('active_model') == 'product.template':
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: # changes start
    # changes end

File: configurator_wizard_in_product/models/product.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            'target': 'new',
+            'res_id': wizard.id,
+        }
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    @api.multi
+    def action_wizard_product_template_configurator(self):
```
Comment (coleste) [inline configurator_wizard_in_product/models/product.py]: rename function...

File: configurator_wizard_in_product/models/product.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    @api.multi
+    def reconfigure_product(self):
+        """ Creates and launches a product configurator wizard with a linked
+        template and variant in order to re-configure a existing product. It is
+        esetially a shortcut to pre-fill configuration data of a variant"""
```
Comment (coleste) [inline configurator_wizard_in_product/models/product.py]: you know...

**Review by coleste — COMMENTED** [review]
and pep8 plz

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        template_products = self.env['product.product'].search(domain)
+        product_found = False
+
+        if self.product_tmpl_id.reuse_variant:
+            for product in template_products:
+                if product.attribute_value_ids == self.value_ids:
+                    product_found = product
+                    break
+
+        if product_found:
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: Bad bad bad!

we are modifying the current variant, not creating / finding another

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        # It should be refactored.
+        # In the meantime, at least make sure that a validation
+        # error legitimately raised in a nested routine
+        # is passed through.
+        domain = [['product_tmpl_id', '=', self.product_tmpl_id.id]]
+
+        template_products = self.env['product.product'].search(domain)
+        product_found = False
+
+        if self.product_tmpl_id.reuse_variant:
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: Reuse_variant should have no impact here.
We are modifying the current variant, not creating / finding another

Your active_model test is way too late...

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                    product_found = product
+                    break
+
+        if product_found:
+            variant = product_found
+            price = product_found.standard_price
+            uom = product_found.uom_id.id
+        else:
+            try:
+                variant = self.product_tmpl_id.create_variant(
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: should never happen... (for tmpl / variant)

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                uom = self.product_tmpl_id.uom_id.id
+            except ValidationError:
+                raise
+            except:
+                raise ValidationError(
+                    _('Invalid configuration! Please check all '
+                      'required steps and fields.')
+                )
+
+        if self.env.context.get('active_model') == 'purchase.order':
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: yeah, it happens too late... you've already created a new variant

File: configurator_wizard_in_product/wizard/product_configurator.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            order = self.env['purchase.order'].browse(self.env.context.get('active_id'))
+            line_vals = {'product_id': variant.id, 'date_planned': datetime.today().strftime(DEFAULT_SERVER_DATETIME_FORMAT),
+                         'product_uom': uom, 'price_unit': price, 'product_qty': 1}
+        # Changes start
+        elif self.env.context.get('active_model') == 'product.template':
```
Comment (coleste) [inline configurator_wizard_in_product/wizard/product_configurator.py]: yeah, it happens too late... you've already created a new variant

File: configurator_wizard_in_product/views/product_views.xml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    <record id="product_normal_form_view" model="ir.ui.view">
+        <field name="name">pproduct.product.form</field>
+        <field name="model">product.product</field>
+        <field name="inherit_id" ref="product.product_normal_form_view"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='attribute_value_ids']" position="before">
+                <button name="reconfigure_product_variant" groups="product_configurator.group_product_configurator"
+                        help="Reconfigure"
+                        icon="fa-cogs"
+                        type="object"
```
Comment (coleste) [inline configurator_wizard_in_product/views/product_views.xml]: grey background color... not sure how to make it white

and we discussed hiding the value_ids or making it readonly.

**Review by elemire — APPROVED** [review]

**Pull request merged**